ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,525 @@
+        ExFreePool(netevent_event_context);
+        netevent_event_context = NULL;
+    }
+    EBPF_EXT_RETURN_RESULT(result);
+}
+
+static void
+_ebpf_netevent_program_context_destroy(
+    _In_opt_ void* context,
+    _Out_writes_bytes_to_(*data_size_out, *data_size_out) uint8_t* data_out,
+    _Inout_ size_t* data_size_out,
+    _Out_writes_bytes_to_(*context_size_out, *context_size_out) uint8_t* context_out,
+    _Inout_ size_t* context_size_out)
+{
+    EBPF_EXT_LOG_ENTRY();
+
+    netevent_event_md_t* netevent_event_context = (netevent_event_md_t*)context;
+    netevent_event_md_t* netevent_event_context_out = (netevent_event_md_t*)context_out;
+
+    if (!netevent_event_context) {
+        goto Exit;
+    }
+
+    if (context_out != NULL && *context_size_out >= sizeof(netevent_event_md_t)) {
+        // Copy the context to the caller.
+        memcpy(netevent_event_context_out, netevent_event_context, sizeof(netevent_event_md_t));
+        *context_size_out = sizeof(netevent_event_md_t);
+
+        // Zero out the event context info.
+        netevent_event_context_out->event_data_start = 0;
+        netevent_event_context_out->event_data_end = 0;
+        *context_size_out = sizeof(netevent_event_md_t);
+    } else {
+        *context_size_out = 0;
+    }
+
+    // Copy the event data to 'data_out'.
+    if (data_out != NULL &&
+        *data_size_out >= (size_t)(netevent_event_context->event_data_end - netevent_event_context->event_data_start)) {
+        memcpy(
+            data_out,
+            netevent_event_context->event_data_start,
+            netevent_event_context->event_data_end - netevent_event_context->event_data_start + 1);
+        *data_size_out = netevent_event_context->event_data_end - netevent_event_context->event_data_start;
+    } else {
+        *data_size_out = 0;
+    }
+
+    ExFreePool(netevent_event_context);
+
+Exit:
+    EBPF_EXT_LOG_EXIT();
+}
+
+//
+// Event Hook NPI client helper functions (invoked by NetEvent as the NPI provider).
+//
+typedef struct _netevent_event_notify_context
+{
+    netevent_event_md_t netevent_event_md;
+} netevent_event_notify_context_t;
+
+void
+_ebpf_netevent_push_event(_In_ netevent_event_md_t* netevent_event)
+{
+    // Logging may delay the event processing, consider enabling only for debugging or if the calling frequency for a
+    // specific use case is low.
+    // EBPF_EXT_LOG_ENTRY();
+
+    if (netevent_event == NULL) {
+        return;
+    }
+
+    ebpf_result_t result;
+    ebpf_extension_hook_client_t* client_context = NULL;
+    netevent_event_notify_context_t netevent_event_notify_context = {0};
+    uint64_t event_size = netevent_event->event_data_end - netevent_event->event_data_start;
+
+    // Currently, the verifier does not support read-only contexts, so we need to copy the event data, rather than
+    // directly passing the existing pointers.
+    // Verifier feature proposal: https://github.com/vbpf/ebpf-verifier/issues/639
```

Comment:
Thanks for creating this issue.
Instead of copying the event data to allocated pool, can it be copied to a buffer on stack? If the supplied event is bigger than the stack buffer, only in that case you allocate. So that way, in majority of the case, you can save the cost of allocation.
Thoughts?

Reply:
Agree, having a dual behavior for the time being that the issue is open is a good idea. I will look into how implementing this looks like.